### PR TITLE
[FeatureStore] custom feature sources

### PR DESCRIPTION
**Pull request by runhli**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py [hunk 1]
```diff
@@ -4,21 +4,50 @@
...
+        path: str = None,
         source_delay: DelayMetadata = None,
-        **kwargs
+        source_process_code: SourceProcessCodeMetadata = None,
+        dict: Dict = None,
+        **kwargs,
     ):
+        if type != "custom":
+            if not (path and not dict and not source_process_code):
+                msg = f"Only path should be provided when source type is {type}"
```
Comment (qjxu) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py:30]: This message is accurate. User might also provide source_delay and time stamp column.

cannot provide source_process_code or kwargs for custom feature source.

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py [hunk 2]
```diff
@@ -4,21 +4,50 @@
...
+                raise ValidationException(
+                    message=msg,
+                    no_personal_data_message=msg,
+                    error_type=ValidationErrorType.INVALID_VALUE,
+                    target=ErrorTarget.FEATURE_SET,
+                    error_category=ErrorCategory.USER_ERROR,
+                )
+        else:
+            if not (dict and source_process_code and not path):
+                msg = f"Only kwargs and source_process_code should be provided when source type is {type}"
```
Comment (qjxu) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py:40]: You cannot provide path for custom feature source.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py [hunk 3]
```diff
@@ -289,13 +293,15 @@ def begin_update(
                         "any data that was materialized "
                         "earlier will not be available. You have to run backfill again."
                     )
-            elif not materialization_identity:
```
Comment (qjxu) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py:292]: why this is removed?

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_feature_set/source_metadata_schema.py [hunk 4]
```diff
@@ -11,14 +11,17 @@
...
 from .timestamp_column_metadata_schema import TimestampColumnMetadataSchema
 
 
 class SourceMetadataSchema(metaclass=PatchedSchemaMeta):
     type = fields.Str(required=True)
-    path = fields.Str(required=True)
+    path = fields.Str(required=False)
     timestamp_column = fields.Nested(TimestampColumnMetadataSchema, required=True)
     source_delay = fields.Nested(DelayMetadataSchema, required=False)
+    source_process_code = fields.Nested(SourceProcessCodeSchema, load_only=True, required=False)
```
Comment (bastrik) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_feature_set/source_metadata_schema.py:23]: If set to load_only, these will not be visible to user if they try to print(featureset)?

Comment (runhli) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/_feature_set/source_metadata_schema.py:23]: Even for now, user will not see feature set spec when print(featureset). User can print feature set spec in data plane, in control plane, the schema is mainly for validation.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py (hunk 3 above)
Comment (runhli) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py:292]: We should throw exceptions whenever offline store/online store is provided and materialization identity is not provided or does not exist.

**Review by bastrik — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py (hunk 3 above)
Comment (qjxu) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py:292]: Seems the validation logic is moved to line 346.

Comment (runhli) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_feature_store_operations.py:292]: yes, if put it at line 292, this means we will check this materialization identity only when input has offline store and there is an existing offline store, this does not include all the scenarios.

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py (hunk 1 above)
Comment (runhli) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_feature_set/source_metadata.py]: updated

**Review by qjxu — APPROVED** [review]

**Review by singankit — APPROVED** [review]

**Pull request merged**